flask_oidc/__init__.py

```diff
@@ -187,11 +196,115 @@ def init_app(self, app):
+        assert secrets != None, "Problem with loading secrets"
+
+        self.client_secrets = list(secrets.values())[0]
+        secrets_cache = DummySecretsCache(secrets)
+
+        # Initialize oauth2client
+        self.flow = flow_from_clientsecrets(
+            current_app.config['OIDC_CLIENT_SECRETS'],
+            scope=current_app.config['OIDC_SCOPES'],
+            cache=secrets_cache)
+
+        assert isinstance(self.flow, OAuth2WebServerFlow)
+
+        current_app.config['OIDC_VALID_ISSUERS'] = (self.client_secrets.get('issuer') or GOOGLE_ISSUERS)
+
+    def load_secrets(self, provider):
+
+        try:
+            static_secrets = current_app.config.get('OIDC_CLIENT_SECRETS', None)
+            if static_secrets:
+                return _json_loads(open(static_secrets,'r').read())
+        except Exception as e:
+            raise Exception("Error reading secrets: {}".format(str(e)))
+
+        if not provider:
+            raise Exception("No Provider specified")
+
+        try:
+            url = provider.get('base_url')
+
+            if not url.endswith('/'):
+                url += '/'
+
+            url += ".well-known/openid-configuration"
+
+            logger.debug("Loading: {}".format(url))
+
+            provider_info = json.load(
+                urlopen(url)
+            )
+
+        except Exception as e:
+            raise Exception("Can not obtain well known information: {}".format(str(e)))
+
+        for path in ['issuer', 'registration_endpoint', 'authorization_endpoint', 'token_endpoint', 'userinfo_endpoint', 'jwks_uri']:
+            if path in provider_info and provider_info[path].startswith('/'):
+                provider_info[path] = "{}{}".format(provider.get('base_url'), provider_info[path])
+
+        registration = provider.get('registration', None)
+
+        if not registration:
+            try:
+                logger.debug("Dynamic Registration...")
+
+                registration = requests.post(
+                    provider_info['registration_endpoint'],
+                    data = json.dumps({
+                        "redirect_uris": self._oidc_callback,
+                        "grant_types": "authorization_code",
+                        "client_name": provider.get('client_name', "Dynamic Registration"),
+                        "response_types": "code",
+                        "token_endpoint_auth_method": "client_secret_post",
+                        "application_type": "native"
+                    }),
+                    headers = {
+                        'Content-Type': "application/json",
+                        'Cache-Control': "no-cache"
+                    }
+                ).json()
+
+                logger.debug("Registration: {}".format(registration))
+
+            except Exception as e:
+                raise Exception("Can not make client registration: {}".format(str(e)))
+
+        try:
+            try:
+               jwks_keys = json.load(
```

Comment:
I think that if we fail to load a jwks_uri, we should just error out entirely, and not even continue starting, since JWKS is required for the protocol.